### PR TITLE
Task/test polling endpoints

### PR DESCRIPTION
**Pull request by krisorr**
# What this does
Adds some tests that exercise the retry and polling in buyClient.getShippingRates().

Addresses #123 

@yervantk @sav007 please review

### CODE REVIEW

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        paymentToken = PaymentToken.createCreditCardPaymentToken("paymenttoken");
+
+        final Field retrofitServiceField = CheckoutServiceDefault.class.getDeclaredField("retrofitService");
+        retrofitServiceField.setAccessible(true);
+        retrofitServiceField.set((((BuyClientDefault) buyClient).checkoutService), checkoutRetrofitService);
+    }
+
+    @Test
+    public void testCompletingCheckoutWithNullCheckoutToken() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
```
Comment (sav007) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java]: it's redundant as this is running on one thread

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        latch.await();
+
+        fail("Expected an IllegalArgumentException");
+    }
+
+    @Test
+    public void testCompletingCheckoutWithEmptyCheckoutToken() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
```
Comment (sav007) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java]: The same as ^^^ above

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        latch.await();
+
+        fail("Expected an IllegalArgumentException");
+    }
+
+    @Test
+    public void testCompletingCheckoutWithNullPaymentToken() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
```
Comment (sav007) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java]: The same ^^^

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+        latch.await();
+
+        fail("Expected an NullPointerException");
+    }
+
+
+    @Test
+    public void testCompleteCheckoutWithNonIOError() throws InterruptedException{
+        final Observable<Response<CheckoutWrapper>> response = Observable.error(new RetrofitError(HttpStatus.SC_PRECONDITION_FAILED, "precondition failed"));
```
Comment (sav007) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java]: I believe we have deleted retrofit error

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+public class CompleteCheckoutTest extends ShopifyAndroidTestCase {
+
+    Checkout checkout;
+    CheckoutWrapper checkoutWrapper;
+
+    CheckoutRetrofitService checkoutRetrofitService;
+
+    PaymentToken paymentToken;
+
+    int callCount;
```
Comment (sav007) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java]: needs to have better name, pollingResponseCallCount

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/ShippingRatesTest.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+ */
+@RunWith(AndroidJUnit4.class)
+public class ShippingRatesTest extends ShopifyAndroidTestCase {
+
+    Checkout checkout;
+    CheckoutWrapper checkoutWrapper;
+
+    CheckoutRetrofitService checkoutRetrofitService;
+
+    int callCount;
```
Comment (sav007) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/ShippingRatesTest.java]: The same note as above in prev test case

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/ShippingRatesTest.java [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        retrofitServiceField.setAccessible(true);
+        retrofitServiceField.set((((BuyClientDefault) buyClient).checkoutService), checkoutRetrofitService);
+
+        shippingRatesWrapper = Mockito.mock(ShippingRatesWrapper.class);
+        Mockito.when(shippingRatesWrapper.getContent()).thenReturn(new ArrayList<ShippingRate>());
+    }
+
+    @Test
+    public void testFetchingShippingRatesWithNullCheckoutToken() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
```
Comment (sav007) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/ShippingRatesTest.java]: No need

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java (hunk 1 above)
Comment (krisorr) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java]: agreed, removed them for all the parameter tests here and in the ShippingRatesTest

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java (hunk 5 above)
Comment (krisorr) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/dataprovider/CompleteCheckoutTest.java]: will update

**Comment by sav007** [comment]
LGTM

**Comment by krisorr** [comment]
Buildkite is failing on unrelated tests, Customer tests are hitting throttling limits.  Will be addressed in another PR.

**Pull request merged**